backends/bmv2/synthesizeValidField.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "synthesizeValidField.h"
+
+#include <boost/range/combine.hpp>
+#include <boost/range/irange.hpp>
+
+#include "backends/bmv2/helpers.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+
+namespace BMV2 {
+
+/// Adds a valid bit to all header types.
+class AddValidField final : public Modifier {
+ public:
+    AddValidField() { setName("AddValidField"); }
+ private:
+    bool preorder(IR::Type_Header* header) override {
+        const cstring validField = V1ModelProperties::validField;
+        if (header->getField(validField) != nullptr) return true;
+        auto field = new IR::StructField(validField, IR::Type::Bits::get(1));
+        header->fields.push_back(field);
+        return true;
+    }
+};
+
+/// Replaces calls to isValid() with references to the header's valid bit.
+class RewriteIsValidCalls final : public Transform {
+ public:
+    RewriteIsValidCalls(P4::ReferenceMap* refMap, P4::TypeMap* typeMap)
+            : refMap(refMap), typeMap(typeMap) {
+        CHECK_NULL(refMap); CHECK_NULL(typeMap);
+        setName("RewriteIsValidCalls");
+    }
+
+ private:
+    const IR::Node* postorder(IR::MethodCallExpression* expr) override;
+
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+};
+
+const IR::Node*
+RewriteIsValidCalls::postorder(IR::MethodCallExpression* expr) {
+    auto instance = P4::MethodInstance::resolve(expr, refMap, typeMap);
+    if (!instance->is<P4::BuiltInMethod>()) return expr;
+    auto builtin = instance->to<P4::BuiltInMethod>();
+    if (builtin->name != IR::Type_Header::isValid) return expr;
+
+    // If isValid() is being used as a table key element, it already behaves
```

Comment:
Note that there are other perfectly legal expressions that evaluate to bool that can be used for table keys, so you still have to handle the other cases in the back-end.